### PR TITLE
Return Validation

### PR DESCRIPTION
**Pull request by robinlacey**
**WIP**

**WHAT** 
We need to validate returns before submitting. ValidatorReturn usecase to 
1) Check if Return is valid
2) Return path of invalid field

### CODE REVIEW

File: lib/local_authority/use_case/validate_return.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    valid = JSON::Validator.validate(schema.schema.to_json, return_data)
+    invalid_paths = valid ? [] : get_paths_from_error_messages(schema, return_data)
+    { valid: valid,
+      invalid_paths: invalid_paths }
+  end
+
+  def get_paths_from_error_messages(schema, return_data)
+    fully_validate = JSON::Validator.fully_validate(schema.schema.to_json, return_data)
+    fully_validate.map do |message|
+      json_property_name_regex_capture = %r{([\w\/]*)}
```
Comment (danielburnley) [inline lib/local_authority/use_case/validate_return.rb]: If youre going to do regexes i'd pull the parsing of them out into a separate method to explain what it's actually parsing

Comment (danielburnley) [reply lib/local_authority/use_case/validate_return.rb]: Historically Regexes are a pain to maintain if you didn't write them

File: lib/local_authority/use_case/validate_return.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      message_parse_regex = %r{
+        '\#/
+          #{json_property_name_regex_capture}
+        '
+        .*
+        '
+          #{json_property_name_regex_capture}
+        '
+      }x
+      message.scan(message_parse_regex).map do |match|
```
Comment (danielburnley) [inline lib/local_authority/use_case/validate_return.rb]: Super nesting, probably worth trying to extract some of these out also

File: spec/unit/local_authority/use_case/vaidate_return_spec.rb [hunk 3]
```diff
@@ -0,0 +1,871 @@
...
+            return_value = use_case.execute(type: project_type, return_data: valid_return_data)
+            expect(return_value[:valid]).to eq(false)
+          end
+        end
+      end
+    end
+
+    context 'given a nested array with invalid entries' do
+      context 'example one' do
+        let(:template) do
```
Comment (danielburnley) [inline spec/unit/local_authority/use_case/vaidate_return_spec.rb]: I wonder if there's some way we could make a builder for these schemas - we make so many and they're so hard to read

File: spec/unit/local_authority/use_case/vaidate_return_spec.rb [hunk 4]
```diff
@@ -0,0 +1,871 @@
+# frozen_string_literal: true
+
+describe LocalAuthority::UseCase::ValidateReturn do
```
Comment (danielburnley) [inline spec/unit/local_authority/use_case/vaidate_return_spec.rb]: Your tests seem to follow a repeating structure of:
```
Given X
    Passes the things to gateways
    Given a valid return
        return valid true
    Given an invalid return
        return valid false
        return paths
```
You could look at using `shared_examples` as a way of refactoring some of this, feel free to grab me if you want a runthrough

File: lib/local_authority/use_case/validate_return.rb [hunk 5]
```diff
@@ -14,30 +14,55 @@ def execute(type:, return_data:)
...
-          path.split('/').map do |node|
-            begin
-              Integer(node)
-            rescue ArgumentError
-              node.to_sym
-            end
-          end
-        end.flatten
-      end.first
+    validation_messages = JSON::Validator.fully_validate(
```
Comment (robinlacey) [inline lib/local_authority/use_case/validate_return.rb]: This is much nicer

File: spec/unit/local_authority/use_case/validate_return_nested_spec.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+describe LocalAuthority::UseCase::ValidateReturn do
```
Comment (danielburnley) [inline spec/unit/local_authority/use_case/validate_return_nested_spec.rb]: :eyes:

Comment (robinlacey) [reply spec/unit/local_authority/use_case/validate_return_nested_spec.rb]: 🤦‍♂️

File: spec/web_routes/validate_return_spec.rb [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    context 'with no type' do
+      before do
+        post '/return/validate', { returnData: return_data }.to_json
+      end
+      it 'returns 400' do
+        expect(last_response.status).to eq(400)
+      end
+    end
+
+    context 'with do returnData' do
```
Comment (danielburnley) [inline spec/web_routes/validate_return_spec.rb]: `with no return data`

File: lib/delivery_mechanism/web_routes.rb [hunk 8]
```diff
@@ -125,6 +125,22 @@ def get_hash(request)
       response.status = 200
     end
 
+    post '/return/validate' do
```
Comment (danielburnley) [inline lib/delivery_mechanism/web_routes.rb]: This needs to be behind a guard

**Pull request merged**